### PR TITLE
 Added trigger for performance monitoring

### PR DESCRIPTION
**Pull request by AmitRoushan**
**What this PR does / why we need it**:
Add trigger point for performance collection after create storage

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

**Comment by ThisIsClark** [comment]
@sushanthakumar Could we add the CI to this repo?

File: delfin/drivers/api.py [hunk 1]
```diff
@@ -135,3 +135,8 @@ def collect_array_metrics(self, context, storage_id, interval,
                                                 storage_id=storage_id)
         return driver.collect_array_metrics(context, storage_id,
                                             interval, is_historic)
+
+    def list_resource_metrics(self, context, storage_id,):
+        """List resource metrics """
+        driver = self.driver_manager.get_driver(context, storage_id=storage_id)
+        return driver.list_resource_metrics(context)
```
Comment (ThisIsClark) [inline delfin/drivers/api.py]: No new line at end of file

File: delfin/drivers/driver.py [hunk 2]
```diff
@@ -18,6 +18,8 @@
 
 @six.add_metaclass(abc.ABCMeta)
 class StorageDriver(object):
+    vendor = ""
+    model = ""
```
Comment (ThisIsClark) [inline delfin/drivers/driver.py]: why adding these two var

File: delfin/drivers/fake_storage/__init__.py [hunk 3]
```diff
@@ -384,3 +384,45 @@ def collect_array_metrics(self, ctx, storage_id, interval, is_history):
                 array_metrics.append(m)
 
         return array_metrics
+
+    @staticmethod
+    def list_resource_metrics(context):
+        """
```
Comment (ThisIsClark) [inline delfin/drivers/fake_storage/__init__.py]: Please check the recommended way of comment of PEP8:
https://www.python.org/dev/peps/pep-0008/#documentation-strings

File: delfin/drivers/manager.py [hunk 4]
```diff
@@ -38,6 +44,105 @@ def __init__(self):
         # each of storage systems so that the session between driver
         # and storage system is effectively used.
         self.driver_factory = dict()
+        self.driver_cls_factory = dict()
+        self._register_drivers()
```
Comment (ThisIsClark) [inline delfin/drivers/manager.py]: Why do we need to change the way of find storage driver class

File: delfin/api/v1/storages.py [hunk 5]
```diff
@@ -264,3 +272,13 @@ def _set_synced_if_ok(context, storage_id, resource_count):
         storage['sync_status'] = resource_count * constants.ResourceSync.START
         storage['updated_at'] = current_time
         db.storage_update(context, storage['id'], storage)
+
+
+def _create_performance_monitoring_task(context, storage_id, metric_list):
+    task_val = dict()
+    task_val.update(storage_id=storage_id)
+    task_val.update(args=metric_list)
+    task_val.update(interval=constants.Task.DEFAULT_INTERVAL)
```
Comment (NajmudheenCT) [inline delfin/api/v1/storages.py]: lets take it from a delfin.conf.. we can have new section in delfin.conf for Performance_monitoring conf

File: delfin/common/constants.py [hunk 6]
```diff
@@ -253,3 +253,9 @@ class ResourceSync(object):
     START = 100
     SUCCEED = 100
     FAILED = 101
+
+
+class Task(object):
+    # Task default interval in sec
```
Comment (NajmudheenCT) [inline delfin/common/constants.py]: lets move all these to delfin.conf

File: delfin/api/v1/storages.py [hunk 7]
```diff
@@ -98,6 +98,14 @@ def create(self, req, body):
             # Post registration, trigger alert sync
             self.task_rpcapi.sync_storage_alerts(ctxt, storage['id'],
                                                  query_para=None)
+
+            # Trigger Performance monitoring
```
Comment (NajmudheenCT) [inline delfin/api/v1/storages.py:107]: Should we put it in different try_except , exception in sync operation should not impact perfr egistration

File: delfin/drivers/driver.py (hunk 2 above)
Comment (AmitRoushan) [reply delfin/drivers/driver.py]: The variable are added to enable Driver class to specify vendor and model name

File: delfin/drivers/api.py (hunk 1 above)
Comment (AmitRoushan) [reply delfin/drivers/api.py]: Added new line as per comment
Thanks

File: delfin/drivers/fake_storage/__init__.py (hunk 3 above)
Comment (AmitRoushan) [reply delfin/drivers/fake_storage/__init__.py]: Modified docStrings as per standard

File: delfin/drivers/manager.py (hunk 4 above)
Comment (AmitRoushan) [reply delfin/drivers/manager.py]: Following are the reasons behind it:
1. Currently drivers are configured and vendor and model name is dependent on setup tool configuration. The Driver class are considered as Delfin storage driver if available in ```delfin.storage.drivers``` namespace. The current implementation adds a check over all deriver being ```StorageDriver```
2. extracting out vendor and model name from setup tool may not be good choice as setup tool keys can be any thing and if miss spelled during deployment leads to feature failure. The implementation decouple the dependency over setuptool keys for vendor and model name. Now Deriver class can provide its specifications.

File: delfin/api/v1/storages.py (hunk 7 above)
Comment (AmitRoushan) [reply delfin/api/v1/storages.py:107]: Added try_except block for performance monitoring

File: delfin/common/constants.py (hunk 6 above)
Comment (AmitRoushan) [reply delfin/common/constants.py]: Moved default_interval to delfin.conf

File: delfin/api/v1/storages.py (hunk 5 above)
Comment (AmitRoushan) [reply delfin/api/v1/storages.py]: Modified as per comment

**Comment by AmitRoushan** [comment]
@NajmudheenCT @ThisIsClark 
Added all review comments. PTAL

File: etc/delfin/delfin.conf [hunk 8]
```diff
@@ -12,6 +12,7 @@ db_backend = sqlalchemy
 
 [scheduler]
 config_path = /etc/delfin/scheduler_config.json
```
Comment (NajmudheenCT) [inline etc/delfin/delfin.conf]: remove this and reference to all.. also we need to remove API / metric-config from router and its implementation

File: etc/delfin/delfin.conf [hunk 9]
```diff
@@ -12,6 +12,7 @@ db_backend = sqlalchemy
...
+default_interval = 30
```
Comment (NajmudheenCT) [inline etc/delfin/delfin.conf]: may be name metric_collect_default_interval ? or change the section name as metric_collect

File: etc/delfin/delfin.conf (hunk 8 above)
Comment (AmitRoushan) [reply etc/delfin/delfin.conf]: remove older performance monitoring codes as per suggestion

**Comment by AmitRoushan** [comment]
@NajmudheenCT @sushanthakumar 
PTAL

File: delfin/common/constants.py [hunk 10]
```diff
@@ -288,3 +288,8 @@ class ResourceSync(object):
     START = 100
     SUCCEED = 100
     FAILED = 101
+
+
+class Task(object):
+    PERFORMANCE_TASK_METHOD = "task.driver.performance_monitoring"
```
Comment (AmitRoushan) [inline delfin/common/constants.py]: The name should be as per class name

File: delfin/task_manager/manager.py [hunk 11]
```diff
@@ -32,6 +29,8 @@
 scheduler_opts = [
     cfg.StrOpt('config_path', default='scheduler',
```
Comment (AmitRoushan) [inline delfin/task_manager/manager.py]: pls remove

Comment (AmitRoushan) [reply delfin/task_manager/manager.py]: done

File: delfin/common/constants.py (hunk 10 above)
Comment (AmitRoushan) [reply delfin/common/constants.py]: Need to make changes post scheduler gets added

**Comment by AmitRoushan** [comment]
CI Success:
image

**Comment by AmitRoushan** [comment]
@NajmudheenCT @sushanthakumar @ThisIsClark 
PTAL. Addressed all review comments

File: etc/delfin/delfin.conf (hunk 9 above)
Comment (AmitRoushan) [reply etc/delfin/delfin.conf]: Removed the code. Not need by api process for performance task trigger

File: delfin/api/v1/router.py [hunk 12]
```diff
@@ -46,12 +45,6 @@ def _setup_routes(self, mapper):
                        action="sync_all",
                        conditions={"method": ["POST"]})
 
-        self.resources['performance'] = performance.create_resource()
-        mapper.connect("storages", "/storages/{id}/metrics-config",
```
Comment (sushanthakumar) [inline delfin/api/v1/router.py:50]: There is schema validation file (delfin/api/schemas/perf_collection.py) which need to be removed

File: delfin/api/v1/storages.py [hunk 13]
```diff
@@ -103,6 +101,21 @@ def create(self, req, body):
...
+                context=ctxt, storage_id=storage['id'])
+            # ignore if capabilities is empty
+            if capabilities:
+                _create_performance_monitoring_task(ctxt, storage['id'],
+                                                    capabilities)
+        except Exception as e:
+            # Unexpected error occurred, while performance monitoring.
+            msg = _('Failed to trigger performance monitoring for storage: '
+                    '%(storage)s. Error: %(err)s') % {'storage': storage['id'],
+                                                      'err': e}
```
Comment (sushanthakumar) [inline delfin/api/v1/storages.py]: better to use six.text_type(e) for details of reason

File: delfin/exception.py [hunk 14]
```diff
@@ -309,3 +309,8 @@ class StorageIsSyncing(Invalid):
 class InvalidIpOrPort(DelfinException):
     msg_fmt = _("Invalid ip or port.")
     code = 400
+
+
+class StorageDriverAlert(DelfinException):
```
Comment (sushanthakumar) [inline delfin/exception.py]: Currently not used?

File: delfin/task_manager/manager.py [hunk 15]
```diff
@@ -16,25 +16,13 @@
...
-from delfin.api.v1.performance import PerformanceController
+from delfin import manager
 from delfin.drivers import manager as driver_manager
 from delfin.task_manager.tasks import alerts
 
 LOG = log.getLogger(__name__)
-CONF = cfg.CONF
-
-scheduler_opts = [
-    cfg.StrOpt('config_path', default='scheduler',
```
Comment (sushanthakumar) [inline delfin/task_manager/manager.py:33]: there is class PerformanceCollectionTask in tasks/resources.py which need to be removed

File: delfin/api/v1/storages.py [hunk 16]
```diff
@@ -195,48 +207,6 @@ def _storage_exist(self, context, access_info):
...
-    def _unregister_perf_collection(self, storage_id):
-
-        schedule = config.Scheduler.getInstance()
-
-        # The path of scheduler config file
-        config_file = CONF.scheduler.config_path
-
-        try:
-            # Load the scheduler configuration file
-            data = config.load_json_file(config_file)
```
Comment (NajmudheenCT) [inline delfin/api/v1/storages.py:207]: implementation of this  needs to be removed from config.py

File: delfin/task_manager/manager.py [hunk 17]
```diff
@@ -46,54 +34,6 @@ def __init__(self, service_name=None, *args, **kwargs):
...
-                storage_id = storage.get('id')
-
-                for resource in storage.keys():
-                    if resource == 'id':
-                        continue
-
-                    resource_type = storage.get(resource)
-                    if (resource_type.get('perf_collection') and
-                            resource_type.get('interval') > constants.
-                            SCHEDULING_MIN_INTERVAL):
```
Comment (NajmudheenCT) [inline delfin/task_manager/manager.py:75]: Need to remove this constants from delfin/common/constants.py
RESOURCE_CLASS_TYPE = {'array_polling': 'ArrayPerformanceCollection'}
SCHEDULING_MIN_INTERVAL =

File: delfin/common/constants.py [hunk 18]
```diff
@@ -288,3 +288,11 @@ class ResourceSync(object):
     START = 100
     SUCCEED = 100
     FAILED = 101
+
+
+class Task(object):
+    DEFAULT_TASK_INTERVAL = 30
+    """Default task interval"""
+    PERFORMANCE_TASK_METHOD = "delfin.task_manager.tasks." \
```
Comment (sushanthakumar) [inline delfin/common/constants.py]: u can use the string - delfin.task_manager.tasks.telemetry.PerformanceCollectionTask

**Comment by sushanthakumar** [comment]
LGTM

File: delfin/task_manager/manager.py (hunk 17 above)
Comment (AmitRoushan) [reply delfin/task_manager/manager.py:75]: addressed as per suggestion

File: delfin/api/v1/storages.py (hunk 16 above)
Comment (AmitRoushan) [reply delfin/api/v1/storages.py:207]: removed as per suggestion

File: delfin/common/constants.py (hunk 18 above)
Comment (AmitRoushan) [reply delfin/common/constants.py]: done. Thanks

File: delfin/task_manager/manager.py (hunk 15 above)
Comment (AmitRoushan) [reply delfin/task_manager/manager.py:33]: done. thanks

File: delfin/exception.py (hunk 14 above)
Comment (AmitRoushan) [reply delfin/exception.py]: removed as per suggestion

File: delfin/api/v1/storages.py (hunk 13 above)
Comment (AmitRoushan) [reply delfin/api/v1/storages.py]: updated the code. Thanks

File: delfin/api/v1/router.py (hunk 12 above)
Comment (AmitRoushan) [reply delfin/api/v1/router.py:50]: remove the validation. Thanks

**Review by NajmudheenCT — APPROVED** [review]
LGTM

**Pull request merged**